---
title: "Harmonia axyridis DNA metabarcoding of gut contents"
author: "Marco Benucci"
date: "7 August 2019"
  output:
  pdf_document: default
  word_document: default
---

```{r knitr setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r environment setup, include=F}
## R(v3.5.1 - Win)
#.libPaths(c("D:\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_win", .libPaths("D:\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_win")))

## R(v3.5.1 - OSX)
#.libPaths(c("D:\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_osx", .libPaths("D:\\Dropbox\\PhD Hull\\PhD docs\\Thesis\\R_stats\\rlib3.5_osx")))

## R(v3.5.1 - ubuntu)
.libPaths(c("/home/mb/Dropbox/PhD Hull/PhD docs/Thesis/R_stats/rlib3.5_unix", .libPaths("/home/mb/Dropbox/PhD\ Hull/PhD\ docs/Thesis/R_stats/rlib3.5_unix")))
```

```{r directory, include=FALSE}
dir()
rm(list=ls())
ls()
```

## Setting working directory and packages

The analysis was done on R (R Core Team, 2018) using the following environment and packages.

```{r packages loading, include=F}
pack.list = c("aod","bipartite","boot","cowplot","devtools","dplyr","EcoSimR","ggplot2","ggsignif","gplots","grid","gridExtra","iNEXT","lattice","lme4","lmtest","MASS","mgcv","msa","network","plyr","reshape2","spaa","stringi","tidyr","UpSetR","vegan","zoo")
new.packages = pack.list[!(pack.list %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages, repos="https://cran.rstudio.com")
lapply(pack.list, require, character.only=T)
```


```{r session info}
sessionInfo()
```

DNA metabarcoding data of Harmonia axyridis gut contents were generated in 3 different sequencing runs. 

* Sequencing run from August 2017 contained gut contents of _H. axyridis_ samples from The Lawns and Thwaite (May 2017) sequenced without the use of blocking primers. COIAug17 includes gut contents of Harmonia axyridis (N=50) sequenced in August 2017 WITHOUT using Blocking primers. The samples were collected from `The Lawns` and from `Thwaite Gardens` (Yorkshire) in May 2017.

* COIDvAug18 and COIHaAug18 include gut contents of Harmonia axyridis sequenced WITH Blocking primers. The samples were collected from Yorkshire and Oxfordshire sites in May and October 2017. COIHaAug18 includes community DNA metabarcoding as well.

# 1. Initial quality control
Input files and first overview. We load the 3 different files into three different data frames

```{r input data}
ha.aug17 = read.csv("data/CO1Aug17-merge-forwonly-nonchimera-c97-cov2_id97-by-taxonomy-readcounts.blast.csv", header=T)
ha1.aug18 = read.csv("data/CO1DvAug18-merged-forwonly_nonchimera-c97-cov2-id97_by-taxonomy-readcounts.csv", header=T)
ha2.aug18 = read.csv("data/CO1HaAug18-merged-forwonly_nonchimera_c97-cov2-id97_by-taxonomy-readcounts.csv", header=T)
```

Before exploring the data we need to extract the taxonomic information contained in the last column of each file. We save this information in 3 separate `.csv` files.

```{r taxonomy file, include=T}
## Custom function to format the taxonomy column

format_taxonomy <- function(input_data){
  for (t in as.data.frame(input_data)){
    n = gsub("\\['|\\']", "", t)
    i = gsub("\\'", "", n)
    taxa.split = data.frame(cbind(i))
  }
  return(taxa.split)
}

## First input file
taxa.aug17 = ha.aug17$taxomomy
ha.aug17 = ha.aug17[,!(colnames(ha.aug17)%in% 'taxomomy')]

taxa.split17=format_taxonomy(taxa.aug17)
taxa.split17=data.frame(taxa.split17)

## Second input file
taxa1.aug18 = ha1.aug18$taxomomy
ha1.aug18 = ha1.aug18[,!(colnames(ha1.aug18)%in% 'taxomomy')]

taxa1.split18=format_taxonomy(taxa1.aug18)
taxa1.split18=data.frame(taxa1.split18)


## Third input file
taxa2.aug18 = ha2.aug18$taxomomy
ha2.aug18 = ha2.aug18[,!(colnames(ha2.aug18)%in% 'taxomomy')]

taxa2.split18=format_taxonomy(taxa2.aug18)
taxa2.split18=data.frame(taxa2.split18)

## Writing taxonomy outputs
write.csv(taxa.split17, "R_outputs/CO1HaAug17_taxonomy.csv")
write.csv(taxa1.split18, "R_outputs/CO1Ha1Aug18_taxonomy.csv")
write.csv(taxa2.split18, "R_outputs/CO1Ha2Aug18_taxonomy.csv")
```

Each initial input file was generated as the result of two separate operations. First DNA centroids were assigned against reference database, the reads that resulted unassigned were then assigned against the NCBI nt database. Each operation generated two separate files which were then merged. The following step is to remove potential duplicate taxa records that would prevent R from definying the row names as the taxa names. Because we removed the last column with the taxonomy information, we can use the whole frame; otherwise we would have needed to avoid the last column in each frame.

```{r removing duplicates, include=T}
ha.aug17 = as.data.frame(ha.aug17 %>% group_by(X.OTU.ID) %>% summarise_all(list(sum)))
ha1.aug18 = as.data.frame(ha1.aug18 %>% group_by(X.OTU.ID) %>% summarise_all(list(sum)))
ha2.aug18 = as.data.frame(ha2.aug18 %>% group_by(X.OTU.ID) %>% summarise_all(list(sum)))
```

We can now set up the rownames for each data frame equal to the species names, and we can remove the first column.

```{r data frames rownames, include=F}
## Setting up rownames equal to the OTU ID
rownames(ha.aug17) = ha.aug17$X.OTU.ID
rownames(ha1.aug18) = ha1.aug18$X.OTU.ID
rownames(ha2.aug18) = ha2.aug18$X.OTU.ID

## Removing the OTU ID column
ha.aug17=ha.aug17[,!(colnames(ha.aug17)%in% 'X.OTU.ID')]
ha1.aug18=ha1.aug18[,!(colnames(ha1.aug18)%in% 'X.OTU.ID')]
ha2.aug18=ha2.aug18[,!(colnames(ha2.aug18)%in% 'X.OTU.ID')]
```

# 2. Initial overview

The data were generated from 3 separate sequencing runs, therefore we want to check the read depth across the 3 runs. The first two sequencing runs (structured reflected into the data frames) contain extra samples which we are not interested, while the last data frame contains only samples regarding _Harmonia axyridis_ and doesn't need to be manipulated. We carry over only the plates that include _H. axyridis_ samples, making sure we transfer also all the Positive and Negative samples from each data frame. These will be used to calculate the filter threshold for each frame indipendently further down the analysis.

```{r extracting Harmonia samples, include=T}
## First frame
ha.reads = subset(ha.aug17[,c(333:402)])

## Second frame
ha1.reads = subset(ha1.aug18[,c(1:8,115:117,169:231,324:329,337:339,347:351)])

## Third frame
ha2.reads = ha2.aug18
```

Read depth per sample.

Since the data were generated into three separate sequencing runs, we check the read depth of all the samples across the three separate frames. We first remove all empty samples before calculating mean and standard deviation values across all three separate frames.
Total number of samples were respectively, `ha.reads` (sequencing run from Aug17) Ntot=354; `ha1.reads` (first run from Aug18) Ntot=339; `ha2.reads` (second run from Aug18) Ntot=351.

```{r boxplot read depth}
colnames(ha.reads)
colnames(ha1.reads)
colnames(ha2.reads)

ha.reads = subset(ha.reads, select = colSums(ha.reads) > 0)
ha1.reads = subset(ha1.reads, select = colSums(ha1.reads) > 0)
ha2.reads = subset(ha2.reads, select = colSums(ha2.reads) > 0)

ha.depth = as.data.frame(ha.reads)
ha1.depth = as.data.frame(ha1.reads)
ha2.depth = as.data.frame(ha2.reads)

ha.depth$library = "First"
ha1.depth$library = "Second"
ha2.depth$library = "Third"

ha.depth = melt(ha.depth)
ha.depth = ha.depth %>% group_by(library, variable) %>% summarise_all(list(sum))

ha1.depth = melt(ha1.depth)
ha1.depth = ha1.depth %>% group_by(library, variable) %>% summarise_all(list(sum))

ha2.depth = melt(ha2.depth)
ha2.depth = ha2.depth %>% group_by(library, variable) %>% summarise_all(list(sum))

# Joining both frames, then plotting

full.depth = rbind(ha.depth, ha1.depth, ha2.depth)

ggplot(full.depth, aes(library, value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(color=c("gray30"), alpha=0.5) +
  theme_bw() + labs(x="Data sets", y="Sample read depth (N reads)")
```

```{r mean read depth, include=F}
ha.reads = subset(ha.reads, select=colSums(ha.reads) > 0)
ha1.reads = subset(ha1.reads, select=colSums(ha1.reads) > 0)
ha2.reads = subset(ha2.reads, select=colSums(ha2.reads) > 0)

## Mean values
mean(colSums(ha.reads))
mean(colSums(ha1.reads))
mean(colSums(ha2.reads))

hist(colSums(ha.reads), breaks=15)
hist(colSums(ha1.reads), breaks=15)
hist(colSums(ha2.reads), breaks=15)
```

Considering our normalisation steps during the libraries preparations, we expect the read coverage per sample to be normally distributed in each run. With this in mind we tested for normality and then we tested for differences in the read depth between the different plates. The results show that all the plates were not significantly difference in regards to reads depth: `ha.reads` and `ha1.reads` (t-test: t = 0.62067, df = 142, p-value = 0.5358), `ha.reads` and `ha2.reads` (t-test: t = -0.95017, df = 407, p-value = 0.3426), `ha1.reads` and `ha2.reads` (t-test: t = -1.7813, df = 439, p-value = 0.07555).

```{r reads distribution, include=T, echo=F}
shapiro.test(colSums(ha.reads))
shapiro.test(colSums(ha1.reads))
shapiro.test(colSums(ha2.reads))

t.test(colSums(ha.reads), colSums(ha1.reads), var.equal = T)
t.test(colSums(ha.reads), colSums(ha2.reads), var.equal = T)
t.test(colSums(ha1.reads), colSums(ha2.reads), var.equal =T)
```

Read depth threshold

To remove low coverage samples we decide to retain samples that contained N reads greater than mean-2sd. This was possible for only the data from the run done in August 2017 (in `ha.reads` data frame); while for the other two data frames we could apply only the mean-sd threshold.

```{r read depth filter, include=T}
cov_ha = mean(colSums(ha.reads))-(2*sd(colSums(ha.reads)))
cov_ha1 = mean(colSums(ha1.reads))-sd(colSums(ha1.reads))
cov_ha2 = mean(colSums(ha2.reads))-sd(colSums(ha2.reads))

ha.cover = subset(ha.reads, select = colSums(ha.reads) >= cov_ha)
ha1.cover = subset(ha1.reads, select = colSums(ha1.reads) >= cov_ha1)
ha2.cover = subset(ha2.reads, select = colSums(ha2.reads) >= cov_ha2)
```

After this initial low coverage filter we retained 98.8%, 97.3% and 97.7% of the total number of reads, and respectively 92.8%, 73.9% and 80.2% number of samples.

```{r reads survival boxplot, include=TRUE, echo=FALSE}
ha.retain = as.data.frame(ha.cover)
ha1.retain = as.data.frame(ha1.cover)
ha2.retain = as.data.frame(ha2.cover)

ha.retain$library = "First"
ha1.retain$library = "Second"
ha2.retain$library = "Third"

# Sum of DNA reads by samples to plot the sample read depth

ha.retain = melt(ha.retain)
ha.retain = ha.retain %>% group_by(library, variable) %>% summarise_all(list(sum))

ha1.retain = melt(ha1.retain)
ha1.retain = ha1.retain %>% group_by(library, variable) %>% summarise_all(list(sum))

ha2.retain = melt(ha2.retain)
ha2.retain = ha2.retain %>% group_by(library, variable) %>% summarise_all(list(sum))

# Joining both frames, then plotting

full.retain = rbind(ha.retain, ha1.retain, ha2.retain)

ggplot(full.retain, aes(library, value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(color=c("gray30"), alpha=0.5) +
  theme_bw() + labs(x="Data sets", y="Sample read depth (N reads)")
```

```{r read survival percentage}
(sum(ha.cover)/sum(ha.reads))*100
(sum(ha1.cover)/sum(ha1.reads))*100
(sum(ha2.cover)/sum(ha2.reads))*100

(count(colSums(ha.cover) > 0)[2]/count(colSums(ha.reads) > 0)[2])*100
(count(colSums(ha1.cover) > 0)[2]/count(colSums(ha1.reads) > 0)[2])*100
(count(colSums(ha2.cover) > 0)[2]/count(colSums(ha2.reads) > 0)[2])*100
```

We check the presence of unwanted taxa and high rank assignment into our data frames. In particular `unassigned` reads, and higher taxa such as unknown `Eukaryota`, `Fungi` and `Bacteria` are going to be removed since they won't be informative at all further downstream. To make the process efficient, we first remove all taxa that have no reads at all as a consequence of previous quality control steps, we then check for the presence of unwanted taxa.

```{r remove unwanted taxa}
ha.clean = subset(ha.cover, subset=rowSums(ha.cover) > 0)
ha1.clean = subset(ha1.cover, subset=rowSums(ha1.cover) > 0)
ha2.clean = subset(ha2.cover, subset=rowSums(ha2.cover) > 0)

## First frame
disc.ha = c("unassigned")
ha.clean = ha.clean[!(rownames(ha.clean)%in% disc.ha),]

## Second frame
disc.ha1 = c("Cyanistes_caeruleus","Parus_major","Salmo_salar",
             "Trebouxia_aggregata","unassigned","uncultured_Jaminaea")
ha1.clean = ha1.clean[!(rownames(ha1.clean)%in% disc.ha1),]

## Third frame
disc.ha2 = c("Columbidae","Eukaryota","Melampsora","Penicillium_sclerotiorum",
             "Sus_scrofa","unassigned","uncultured_Jaminaea","Verticillium")
ha2.clean = ha2.clean[!(rownames(ha2.clean)%in% disc.ha2),]
```

We check for the survival of samples and DNA reads

```{r retained reads and samples}
(sum(ha.clean)/sum(ha.reads))*100
(sum(ha1.clean)/sum(ha1.reads))*100
(sum(ha2.clean)/sum(ha2.reads))*100

((sum(ha.cover)/sum(ha.reads))*100-(sum(ha.clean)/sum(ha.reads))*100)
((sum(ha1.cover)/sum(ha1.reads))*100-(sum(ha1.clean)/sum(ha1.reads))*100)
((sum(ha2.cover)/sum(ha2.reads))*100-(sum(ha2.clean)/sum(ha2.reads))*100)
```

We retained approx. 93.6%, 90% and 78.7% of the initial N reads. After having removed the undesired taxa, we can now extract the information from the `Positive` samples, and calculate the contamination thresholds for each data set. We need to extract the positive samples from all frames. In the sequencing run done in August 2017 (`ha.clean` frame) the Positive taxa used was _Osmia bicornis_, while in both sequencing runs from August 2018 (`ha1.clean` and `ha2.clean` frames) the positive taxa used was _Lumbricus rubellus_.

# 3. Calculating contamination threshold

The first data frame containes data from the sequencing run from August 2017, which includes gut content samples of _Harmonia axyridis_ from Lawns (N=25) and Thwaite Gardens (N=25) collected in May 2017 amplified without the use of blocking primers. To calculate the contamination threshold we decided to account for all laboratory contaminations (e.g. Human DNA) in our `POSITIVE` samples. In the first data frame we see that the majority of the contamination in the `POSITIVE` sample comes from _Harmonia axyridis_ (~62% of the reads in the positive samples). We then decided to focus on the levels of other known contaminations, which in this case is _Homo sapiens_.

```{r threshold first frame, include=T}
## Positive samples from `ha.clean` (August 2017)
ha.pos = ha.clean[colnames(ha.clean) == "POSTITIVE"]
ha.pos[,]/colSums(ha.pos)
rownames(ha.pos)

ha.pre.filter = as.data.frame(t(ha.clean))
ha.pre.filter = ha.pre.filter[,]/rowSums(ha.pre.filter)
ha.pre.filter[is.na(ha.pre.filter)] <- 0

thres =  0.001
```

We have to exclude the contamination from _Harmonia axyridis_ as that would be too high for a fixed contamination threshold and we would lose any informations about prey-predator interactions. The remaining contaminations appear to be: _Hymenoptera_ (0.468%), and _Homo sapiens_ (0.0969%). Based on the levels of these contaminations we decided to apply a threshold of 0.1%.

```{r filter first frame, include=T}
ha.post.filter = as.data.frame(ha.pre.filter)
ha.post.filter[][ha.post.filter <= thres] <- 0

ha.clean = as.data.frame(t(ha.clean))
ha.clean[][ha.clean[,]/rowSums(ha.clean) <= thres] <- 0

## Writing pre- and post-filter data to output
write.csv(ha.clean, "R_outputs/HaAug17_cleaned_reads.csv")
write.csv(ha.pre.filter, "R_outputs/HaAug17_pre-filter.csv")
write.csv(ha.post.filter, "R_outputs/HaAug17_post-filter.csv")
```

We add now information regarding sampling sites (extracted from the samples names), use of blocking primers or not, pre- or post-filter, and the sample type (gut contents or community)

```{r adding variables first frame, include=F}
ha.pre = as.data.frame(ha.pre.filter)
ha.post = as.data.frame(ha.post.filter)

ha.pre$type = "pre.filter"
ha.post$type = "post.filter"

ha.clean$sample_id = gsub(".nc.blast$", "", rownames(ha.clean))
ha.pre$sample_id = gsub(".nc.blast$", "", rownames(ha.pre))
ha.post$sample_id = gsub(".nc.blast$", "", rownames(ha.post))

ha.clean$sample_type = "gut"
ha.pre$sample_type = "gut"
ha.post$sample_type = "gut"

ha.clean$pcr = "no-blocking"
ha.pre$pcr = "no-blocking"
ha.post$pcr = "no-blocking"
```

The second data frame includes data from the first sequencing run from August 2018, which contains _H. axyridis_ gut contents samples from Oxfordshire sites collected in May 2017.

```{r threshold second frame, include=TRUE}
## Positive samples Aug18_1
pos18.1 = grep("POS", colnames(ha1.clean))
colSums(ha1.clean[,pos18.1])
ha1.pos = as.data.frame(ha1.clean[,pos18.1])

tot.pos18.1 = sum(ha1.pos)
lr.pos18.1 = sum(ha1.pos[c("Lumbricus_rubellus","Lumbricidae","Harmonia_axyridis"),])
cont18.1 = tot.pos18.1-lr.pos18.1
cont18.1

#thres18.1 = cont18.1/sum(ha1.pos)

ha18.1.pre.filter = as.data.frame(t(ha1.clean))
ha18.1.pre.filter = ha18.1.pre.filter/rowSums(ha18.1.pre.filter)
ha18.1.pre.filter[is.na(ha18.1.pre.filter)] <- 0

max(ha18.1.pre.filter$Homo_sapiens)

thres18.1 = 0.003
```

We calculated the threshold equal to approx. `3.8e-05`; however the max values from known contamination taxa across samples outside of the positive appear higher. In particular we detect contamination of: _Homo sapiens_ = 0.002804504. We decided to apply 0.003 threshold.

```{r filter second frame}
ha18.1.post.filter = as.data.frame(ha18.1.pre.filter)
ha18.1.post.filter[,][ha18.1.post.filter[,] <= thres18.1] <- 0

ha1.clean = as.data.frame(t(ha1.clean))
ha1.clean[][ha1.clean[,]/rowSums(ha1.clean) <= thres18.1] <- 0

## Write output to csv

write.csv(ha1.clean, "R_outputs/HaAug18_1_cleaned_reads.csv")
write.csv(ha18.1.pre.filter, "R_outputs/HaAug18_1_pre-filter.csv")
write.csv(ha18.1.post.filter, "R_outputs/HaAug18_1_post-filter.csv")
```

We add now information regarding sampling sites (extracted from the samples names), use of blocking primers or not, pre- or post-filter, and the sample type (gut contents or community)

```{r adding variable second frame, include=F}
ha18.1.pre = as.data.frame(ha18.1.pre.filter)
ha18.1.post = as.data.frame(ha18.1.post.filter)

ha18.1.pre$type = "pre.filter"
ha18.1.post$type = "post.filter"

ha1.clean$sample_id = gsub("[.]pl.[a-zA-Z0-9.]{3,5}.nc.blast$", "", rownames(ha1.clean))
ha18.1.pre$sample_id = gsub("[.]pl.[a-zA-Z0-9.]{3,5}.nc.blast$", "", rownames(ha18.1.pre))
ha18.1.post$sample_id = gsub("[.]pl.[a-zA-Z0-9.]{3,5}.nc.blast$", "", rownames(ha18.1.post))

ha1.clean$sample_type = "gut"
ha18.1.pre$sample_type = "gut"
ha18.1.post$sample_type = "gut"

ha1.clean$pcr = "with blocking"
ha18.1.pre$pcr = "with blocking"
ha18.1.post$pcr = "with blocking"
```

The last data frame includes data from the second from August 2018 containing _H. axyridis_ contents samples from Yorkshire sites collected in May 2017, plus the Oxfordshire and Yorkshire sites collected in October 2017. This frame contains also all the community tissue samples.

```{r threshold third frame, include=FALSE}
## Positive samples Aug18_2
pos18.2 = grep("POS", colnames(ha2.clean))
colSums(ha2.clean[,pos18.2])
ha2.pos = as.data.frame(ha2.clean[,pos18.2])
ha2.pos = subset(ha2.pos, subset=rowSums(ha2.pos) > 0)
rowSums(ha2.pos)

tot.pos18.2 = sum(ha2.pos)
lr.pos18.2 = sum(ha2.pos[c("Lumbricus_rubellus","Lumbricidae","Harmonia_axyridis"),])
cont18.2 = tot.pos18.2-lr.pos18.2
cont18.2

ha18.2.pre.filter = as.data.frame(t(ha2.clean))
ha18.2.pre.filter = ha18.2.pre.filter/rowSums(ha18.2.pre.filter)
ha18.2.pre.filter[is.na(ha18.2.pre.filter)] <- 0

thres18.2 =  0.001

max(ha18.2.pre.filter$Dikerogammarus_villosus)
max(ha18.2.pre.filter$Homo_sapiens)
```

We didn't detect any other contamination in the positive samples outside of _H. axyridis_ DNA; however known contaminations and their max value of detection in the actual samples were: _Dikerogammarus villosus_ = 0.0005219752, and _Homo sapiens_ = 0.0006574622. We then decided to put the threshold equal to `0.0006` to remove these main contaminations, despite retaining some Human contamination.

```{r filter third frame}
ha18.2.post.filter = as.data.frame(ha18.2.pre.filter)
ha18.2.post.filter[][ha18.2.post.filter <= thres18.2] <- 0

ha2.clean = as.data.frame(t(ha2.clean))
ha2.clean[][ha2.clean[,]/rowSums(ha2.clean) <= thres18.2] <- 0

## Write output to csv
write.csv(ha2.clean, "R_outputs/HaAug18_2_cleaned_reads.csv")
write.csv(ha18.2.pre.filter, "R_outputs/HaAug18_2_pre-filter.csv")
write.csv(ha18.2.post.filter, "R_outputs/HaAug18_2_post-filter.csv")
```

We add now information regarding sampling sites (extracted from the samples names), use of blocking primers or not, pre- or post-filter, and the sample type (gut contents or community)

```{r adding variable third frame, include=FALSE}
ha18.2.pre = as.data.frame(ha18.2.pre.filter)
ha18.2.post = as.data.frame(ha18.2.post.filter)

ha18.2.pre$type = "pre.filter"
ha18.2.post$type = "post.filter"

ha2.clean$sample_id = gsub("[.]pl.[a-zA-Z0-9.]{3,5}.nc.blast$", "", rownames(ha2.clean))
ha18.2.pre$sample_id = gsub("[.]pl.[a-zA-Z0-9.]{3,5}.nc.blast$", "", rownames(ha18.2.pre))
ha18.2.post$sample_id = gsub("[.]pl.[a-zA-Z0-9.]{3,5}.nc.blast$", "", rownames(ha18.2.post))

ha2.clean$sample_type = "NA"
ha2.clean[grep("Ha", ha2.clean$sample_id), "sample_type"] = "gut"
ha2.clean[-grep("Ha", ha2.clean$sample_id), "sample_type"] = "community"

ha18.2.pre$sample_type = "NA"
ha18.2.pre[grep("Ha", ha18.2.pre$sample_id),"sample_type"] = "gut"
ha18.2.pre[-grep("Ha", ha18.2.pre$sample_id),"sample_type"] = "community"

ha18.2.post$sample_type = "NA"
ha18.2.post[grep("Ha", ha18.2.post$sample_id),"sample_type"] = "gut"
ha18.2.post[-grep("Ha", ha18.2.post$sample_id),"sample_type"] = "community"

ha2.clean$pcr = "with blocking"
ha18.2.pre$pcr = "with blocking"
ha18.2.post$pcr = "with blocking"
```

# 4. Blocking primers
We test here the effects of using blocking primers in gut contents analysis of _Harmonia axyridis_ in the UK. In our case we developed Harmonia-specific blocking primers with a 3'-3C tail. These were added during the first PCR of both the sequencing runs done in August 2018. No blocking primer was added in the sequencing run from August 2017. In total we have same N=50 samples from May 2017 collected in Lawns and Thwaite Gardens that were sequenced twice in two separate runs, one without and one with blocking primers. First we extract the samples from their respective data frames, then we investigate the difference across these 50 samples that were sequenced with and without blocking primers (Ntot = 100).

```{r boxplot blk primers, include=TRUE, echo=FALSE}
no.blk.reads = as.data.frame(ha.clean)
blk.reads = as.data.frame(ha2.clean)

keep = grep("LAWN|THWA", no.blk.reads$sample_id)
no.blk.reads = no.blk.reads[keep,]
no.blk.reads$time = "May"
no.blk.reads$site = gsub("[.][a-z]{3}.[A-Za-z0-9]{3,4}$","", no.blk.reads$sample_id)

blk.reads$time = "NA"
blk.reads[grep("may", rownames(blk.reads)),"time"] = "May"
blk.reads[grep("oct", rownames(blk.reads)),"time"] = "Oct"
blk.reads$site = gsub("[.][a-z]{3}(.[A-Za-z0-9]{3,4})*$","", blk.reads$sample_id)

keep = grep("LAWN|THWA", blk.reads$sample_id)
blk.reads = blk.reads[keep,]
rownames(blk.reads) = blk.reads$sample_id
blk.reads = blk.reads[][blk.reads$sample_type == "gut",]
blk.reads = blk.reads[][blk.reads$time == "May",]

## Creating and calculating the ratio in data frames

no.blk.ratio = as.data.frame(no.blk.reads)
blk.ratio = as.data.frame(blk.reads)

no.blk.ratio = subset(no.blk.ratio[,1:10], select = colSums(no.blk.ratio[,1:10]) > 0)
no.blk.ratio = no.blk.ratio/rowSums(no.blk.ratio)

blk.ratio = subset(blk.ratio[,1:143], select = colSums(blk.ratio[,1:143]) > 0)
blk.ratio = blk.ratio/rowSums(blk.ratio)

no.blk.ratio = cbind(no.blk.ratio, no.blk.reads[,11:15])
blk.ratio = cbind(blk.ratio, blk.reads[,144:148])

## Removing the empty taxa in the frames with the reads count

no.blk.reads = subset(no.blk.reads[,1:10], select = colSums(no.blk.reads[,1:10]) > 0)
blk.reads = subset(blk.reads[,1:143], select = colSums(blk.reads[,1:143]) > 0)

no.blk.reads = cbind(no.blk.reads, no.blk.ratio[,5:9])
blk.reads = cbind(blk.reads, blk.ratio[,6:10])

## Removing the remaining contamination of _Lumbricus rubellus_ (the positive taxa) in the frame with blocking primers.
blk.reads = blk.reads[,!(colnames(blk.reads) %in% "Lumbricus_rubellus")]
blk.ratio = blk.ratio[,!(colnames(blk.ratio) %in% "Lumbricus_rubellus")]
```

Before testing whether the blocking primers had an impact on the DNA reads of prey and predator, we needed to make sure that both subsets were comparable in terms of coverage. The Wilcoxon test showed that the coverage of the two subsets were significanlty different (Wilcoxon test: W = 1492, p-value = 0.0063). We decided therefore to include a normalisation step by dividing DNA reads by the total number of reads in each set, and multiplying by 1e06 (Wilcoxon test on normalised reads: W = 1144, p-value = 0.8903).

```{r normalisation, include=F}
wilcox.test(rowSums(no.blk.reads[,1:4]), rowSums(blk.reads[,1:4]))

norm.no.blk.reads = (no.blk.reads[,1:4]/sum(no.blk.reads[,1:4]))*1000000
norm.blk.reads = (blk.reads[,1:4]/sum(blk.reads[,1:4]))*1000000

wilcox.test(rowSums(norm.no.blk.reads), rowSums(norm.blk.reads))
```

After all the previous quality controls and contamination controls we retained the great majority of the gut content samples for the sequencing run without blocking primers (all N=50 samples retained), and with blocking primers (N=45 out of N=50 initial samples). We could now test whether there was an association between the use of blocking primers and the number of reads that we obtained for the prey items and the predator, and we obtain a significant association between the presence of the blocking primers and the number of reads obtained from the sequencing (Chi-squared test: Chi-squared = 13956, df = 1, p-value < 2.2e-16)

```{r chi-squared on normalised, include=T, echo=F}
prey.no.blk = sum(norm.no.blk.reads[,colnames(norm.no.blk.reads) != "Harmonia_axyridis"])
pred.no.blk = sum(norm.no.blk.reads$Harmonia_axyridis)

prey.blk = sum(norm.blk.reads[,colnames(norm.blk.reads) != "Harmonia_axyridis"])
pred.blk = sum(norm.blk.reads$Harmonia_axyridis)

chisq.test(rbind(c(prey.blk, prey.no.blk), c(pred.blk, pred.no.blk)))
```

```{r blk vs no-blk plot}
plot_no_blk = as.data.frame(norm.no.blk.reads[,1:4])
plot_blk = as.data.frame(norm.blk.reads[,1:4])

plot_no_blk = as.data.frame(t(plot_no_blk))
plot_blk = as.data.frame(t(plot_blk))

plot_no_blk$pcr = "no-blocking primers"
plot_blk$pcr = "blocking primers"

plot_no_blk$specie_type = "NA"
plot_no_blk[grep("Harmonia_axyridis", rownames(plot_no_blk)),"specie_type"] = "predator"
plot_no_blk[grep("Harmonia_axyridis", rownames(plot_no_blk), invert = T), "specie_type"] = "prey"

plot_blk$specie_type = "NA"
plot_blk[grep("Harmonia_axyridis", rownames(plot_blk)),"specie_type"] = "predator"
plot_blk[grep("Harmonia_axyridis", rownames(plot_blk), invert = T), "specie_type"] = "prey"

plot_no_blk$specie = rownames(plot_no_blk)
plot_blk$specie = rownames(plot_blk)

plot_no_blk[plot_no_blk == 0] <- NA
plot_blk[plot_blk == 0] <- NA

plot_frame = full_join(plot_blk, plot_no_blk)

# We set the undetected taxa in either frame equal to "NA".

plot_frame["9",] = plot_frame[plot_frame$specie == "Pyrrhalta_viburni",]
plot_frame["9","pcr"] = "blocking primers"
plot_frame["9",!c(colnames(plot_frame) %in% c("pcr","specie","specie_type"))] <- NA

plot_frame["10",] = plot_frame[plot_frame$specie == "Drepanosiphum_platanoidis",]
plot_frame["10","pcr"] = "no-blocking primers"
plot_frame["10",!c(colnames(plot_frame) %in% c("pcr","specie","specie_type"))] <- NA

full_blk_melt = melt(plot_frame)
```

```{r blk vs no-blk boxplot}
full_blk_melt$pcr = factor(full_blk_melt$pcr, levels=c("no-blocking primers","blocking primers"))
box_a = ggplot(full_blk_melt, aes(specie_type, value)) +
  geom_boxplot(aes(fill=specie_type), outlier.shape = NA) +
  geom_jitter(alpha=0.5, size=4) +
  labs(x="", y="Normalised DNA reads") +
  theme_bw() +
  facet_grid(.~pcr) +
  theme(legend.position="none", axis.text.x=element_blank(), axis.ticks.x=element_blank(), axis.text=element_text(size=15), axis.title=element_text(size=15), strip.text=element_text(size=15), plot.margin=unit(c(0,0.25,0,0.25), unit="cm"))

full_blk_melt$pcr = factor(full_blk_melt$pcr, levels=c("no-blocking primers","blocking primers"))
box_b = ggplot(full_blk_melt,(aes(specie_type, log10(value)))) +
  geom_boxplot(aes(fill=specie_type), outlier.shape = NA) +
  geom_jitter(alpha=0.5, size=4) +
  labs(x="", y="Log10(Normalised DNA reads)") +
  theme_bw() +
  facet_grid(.~pcr) +
  theme(legend.position="right", legend.title=element_blank(), legend.text=element_text(size=15), axis.text.x=element_blank(), axis.ticks.x=element_blank(), axis.text=element_text(size=15), axis.title=element_text(size=15), strip.text=element_text(size=15), plot.margin=unit(c(0,0,0,0.25), unit="cm"))

png("R_images/boxplot_blk_vs_no-blk.png", height=4000, width=4000, units="px", res=300)
plot_grid(box_a, box_b, labels=c("a)","b)"), ncol=2, nrow=1)
dev.off()
```

```{r blk vs no-blk heatmap}
heat_blk_melt = full_blk_melt
heat_blk_melt[is.na(heat_blk_melt)] = 0
heat_blk_melt[heat_blk_melt$variable == "THWA.may.Ha23" & heat_blk_melt$pcr == "blocking primers", "value"] = NA
heat_blk_melt[heat_blk_melt$variable == "THWA.may.Ha7" & heat_blk_melt$pcr == "blocking primers", "value"] = NA
heat_blk_melt[heat_blk_melt$variable == "LAWN.may.Ha22" & heat_blk_melt$pcr == "blocking primers", "value"] = NA
heat_blk_melt[heat_blk_melt$variable == "LAWN.may.Ha23" & heat_blk_melt$pcr == "blocking primers", "value"] = NA
heat_blk_melt[heat_blk_melt$variable == "LAWN.may.Ha24" & heat_blk_melt$pcr == "blocking primers", "value"] = NA
```

```{r heatmap plot}
a<-ggplot(heat_blk_melt[heat_blk_melt$specie %in% c("Drepanosiphum_platanoidis","Pyrrhalta_viburni"),],(aes(pcr, variable))) +
  geom_tile(aes(fill=value), linetype=1, color="gray70", size=0.2) +
  scale_fill_gradientn(colors=c("white","grey87","orange","blue"), values=c(0,0.01,0.5,1), na.value="black", name="Normalised\nDNA reads\n") + 
  theme_classic() +
  facet_grid(.~specie) + labs(x="", y="")

b<-ggplot(heat_blk_melt[heat_blk_melt$specie %in% c("Chrysomelidae","Euceraphis_betulae"),],(aes(pcr, variable))) +
  geom_tile(aes(fill=value), linetype=1, color="gray70", size=0.2) +
  scale_fill_gradientn(colors=c("white","grey87","orange","blue"), values=c(0,0.01,0.5,1), na.value="black", name="Normalised\nDNA reads\n") + 
  theme_classic() +
  facet_grid(.~specie) + labs(x="", y="")

c<-ggplot(heat_blk_melt[heat_blk_melt$specie_type %in% "predator",],(aes(pcr, variable))) +
  geom_tile(aes(fill=value), linetype=1, color="gray70", size=0.2) +
  scale_fill_gradientn(colors=c("grey87","orange","blue"), values=c(0,0.5,1), na.value="black", name="Normalised\nDNA reads\n") +  
  theme_classic() +
  facet_grid(.~specie) + labs(x="", y="")

png("R_images/heatmap_blk_vs_no-blk.png", height=3000, width=6000, units="px", res=300)
grid.arrange(a,b,c, ncol=3, nrow=1, left="Sample ID")
dev.off()
```

# 5. Detection of prey taxa

We want to investigate now the temporal and spatial networks we could detectd from _Harmonia axyridis_ gut contents. At this stage we will use only the datasets that were sequenced with blocking primers, and to better investigate the prey DNA we detected, we decided to remove the predator from our datasets. 
We start by exploring the data splitting them by sampling area (Oxfordshire or Yorkshire), by sampling time (May or October), and by sample type (gut contents or community).

```{r adding info to data frames, echo=FALSE}
ha1.clean$site = gsub("[.][a-z]{3}(.[A-Za-z0-9]{3,4})*$","", ha1.clean$sample_id)
ha1.clean$time = "May"

ha2.clean$site = gsub("[.][a-z]{3}(.[A-Za-z0-9]{3,4})*$","", ha2.clean$sample_id)
ha2.clean$time = "NA"
ha2.clean[grep("may", rownames(ha2.clean)),"time"] = "May"
ha2.clean[grep("oct", rownames(ha2.clean)), "time"] = "Oct"
```

```{r merging harmonia frames, include=F}
ha.full = full_join(ha1.clean, ha2.clean)

ha.full = ha.full[,][,!c(colnames(ha.full) %in% "Lumbricus_rubellus")]
pn = grep("POS|NEG", ha.full$sample_id)
ha.full = ha.full[-pn,]
```


```{r Oxfordshire & Yorkshire sites}
oxf = grep("AoT|KID|RAIL|STRE|WALL|SUTT", ha.full$sample_id, ignore.case = T)
oxf.full = ha.full[oxf,]

ha.ox = grep("Ha", oxf.full$sample_id)
oxf.comms = oxf.full[-ha.ox,]
oxf.comms$county = "Oxfordshire"
oxf.guts = oxf.full[ha.ox,]
oxf.guts$county = "Oxfordshire"

yor = grep("LAWN|THWA|PEAR|OAK|BEV|UNI", ha.full$sample_id, ignore.case = T)
yor.full = ha.full[yor,]

ha.yor = grep("Ha", yor.full$sample_id)
yor.comms = yor.full[-ha.yor,]
yor.comms$county = "Yorkshire"
yor.guts = yor.full[ha.yor,]
yor.guts$county = "Yorkshire"
```

```{r testing difference of prey taxa in guts, include=F}
oxf.guts[is.na(oxf.guts)] <- 0
infos = oxf.guts[,colnames(oxf.guts) %in% c("sample_id","sample_type","pcr","site","time","county")]

# We select only the taxa with N reads great than zero

ha.oxf = subset(oxf.guts[,!c(colnames(oxf.guts) %in% colnames(infos))], select = colSums(oxf.guts[,!c(colnames(oxf.guts) %in% colnames(infos))]) > 0)
ha.oxf = cbind(ha.oxf, infos)
prey.oxf = ha.oxf[,!(colnames(ha.oxf) %in% c("Harmonia_axyridis","Homo_sapiens"))]

prey.oxf[,!c(colnames(prey.oxf) %in% c("sample_id","sample_type","pcr","site","time","county"))][prey.oxf[,!c(colnames(prey.oxf) %in% c("sample_id","sample_type","pcr","site","time","county"))] > 0] <- 1

# We do the same process for the Yorkshire samples

yor.guts[is.na(yor.guts)] <- 0
infos = yor.guts[colnames(yor.guts) %in% c("sample_id","sample_type","pcr","site","time","county")]

# Again we select only the taxa with N reads great than zero

ha.yor = subset(yor.guts[,!c(colnames(yor.guts) %in% colnames(infos))], select = colSums(yor.guts[,!c(colnames(yor.guts) %in% colnames(infos))]) > 0)
ha.yor = cbind(ha.yor, infos)
prey.yor = ha.yor[,!(colnames(ha.yor) %in% c("Harmonia_axyridis","Homo_sapiens"))]

prey.yor[,!c(colnames(prey.yor) %in% c("sample_id","sample_type","pcr","site","time","county"))][prey.yor[,!c(colnames(prey.yor) %in% c("sample_id","sample_type","pcr","site","time","county"))] > 0] <- 1
```

```{r}
oxf.guts$sample_id
```


```{r join full guts samples, include=F}
full.guts = full_join(prey.oxf, prey.yor)

num.taxa = as.data.frame(full.guts[,!c(colnames(full.guts) %in% c("sample_id","sample_type","pcr","site"))] %>% group_by(time, county) %>% summarise_all(list(sum)))
num.taxa[is.na(num.taxa)] <- 0
num.taxa[,!c(colnames(num.taxa) %in% c("time","county"))][num.taxa[,!c(colnames(num.taxa) %in% c("time","county"))] > 0] <- 1

colnames(full.guts)
colnames(num.taxa)

## Removing taxa at Order level

full.guts$Dinocampus_coccinellae = rowSums(full.guts[,c(colnames(full.guts) %in% c("Dinocampus","Dinocampus_coccinellae"))])
num.taxa$Dinocampus_coccinellae = rowSums(num.taxa[,c(colnames(num.taxa) %in% c("Dinocampus","Dinocampus_coccinellae"))])

full.guts = full.guts[,!c(colnames(full.guts) %in% c("Psocoptera","Hemiptera","Dinocampus"))]
num.taxa = num.taxa[,!c(colnames(num.taxa) %in% c("Psocoptera","Hemiptera","Dinocampus"))]

full.guts = melt(full.guts)
num.taxa = melt(num.taxa)
```

We detected a total of 49 interactions in Oxfordshire in May (N tot=63), only 10 interactions in Oxfordshire in October (N tot=59); while we detected 28 and 22 interactions respectively in Yorkshire in May (N tot=104) and October (N tot=99).

```{r guts occurrences plot, include=T, echo=F}
png("R_images/Occurr_taxa_in_guts.png", width=4000, height=4000, units="px", res=300)
ggplot(full.guts, aes(time,value)) +
  geom_bar(stat="identity", position="stack") +
  facet_grid(.~county) +
  coord_cartesian(ylim=c(0,30)) +
  scale_y_continuous(breaks=seq(0,30,5)) +
  labs(title="Total occurrence of taxa in Harmonia axyridis gut contents\nacross counties and seasons", y="N occurrences of taxa", x="Sampling time")
dev.off()
```

The individual _H. axyridis_ show a low diversity of prey utilised. Individuals from Oxfordshire in May showed a higher number of prey taxa detected (N prey taxa=10), while in comparison the same sites but from October showed the least number of prey detected (N prey=2 out of 59 individuals sampled). These results from October match our expecteation considering the mostly empty guts observed during dissections.
Yorskhire samples show as well a low prey diversity for both May (N prey=4 out of 104 sampled), and October (N prey=7 out of 99 individuals sampled).

```{r barplot N prey taxa}
png("R_images/N_taxa_in_guts.png", width=4000, height=4000, units="px", res=300)
ggplot(num.taxa, aes(time,value)) +
  geom_bar(stat="identity", position="stack") +
  facet_grid(.~county) +
  scale_y_continuous(breaks = seq(0,10,2)) +
  labs(title="N taxa detected in Harmonia axyridis gut contents across counties and seasons", y="N taxa", x="Sampling time")
dev.off()
```


```{r join reads.guts, include=T, echo=F}
reads.guts = full_join(ha.oxf, ha.yor)
reads.guts[is.na(reads.guts)] <- 0

## merging taxa

reads.guts$Dinocampus_coccinellae = rowSums(reads.guts[,c(colnames(reads.guts) %in% c("Dinocampus","Dinocampus_coccinellae"))])

reads.guts = reads.guts[,!c(colnames(reads.guts) %in% "Dinocampus")]

reads.guts$Euceraphis = rowSums(reads.guts[,c(colnames(reads.guts) %in% c("Euceraphis","Euceraphis_betulae"))])
reads.guts = reads.guts[,!c(colnames(reads.guts) %in% "Euceraphis_betulae")]
```

```{r read.guts plot, include=T, echo=F}
reads.guts = reads.guts[,!(colnames(reads.guts) %in% c("Harmonia_axyridis","Homo_sapiens","Hemiptera"))]
reads.guts[is.na(reads.guts)] <- 0
melt.guts = melt(reads.guts)
melt.guts = melt.guts[melt.guts$value > 0,]

color16=c("#E69F00","#56B4E9","#CC79A7","#009E73","#F0E442","#999999","#0072B2","#D55E00",
          "#000000","#993300","#9ACD32","#FF0000","#33FF33","#6633CC","#68228B","#0000FF")
png("R_images/guts_composition_all_ha.png", width=4000, height=4000, units="px", res=300)
ggplot(melt.guts, aes(sample_id, value)) +
  geom_bar(stat="identity", position="fill", aes(fill=variable)) +
  coord_flip() +
  scale_fill_manual(values = color16, name = "") +
  facet_grid(county*time~., scales="free") + 
  theme(legend.position = "bottom") +
  labs(title="Gut content composition from read count data for all individuals\n(Harmonia DNA excluded)") +
  guides(fill=guide_legend(nrow=7, byrow=T))
dev.off()
```

```{r read.guts plot 2}
png("R_images/guts_composition_across_oxf_york.png", width=4000, height=4000, units="px", res=300)
ggplot(melt.guts, aes(time, value)) +
  geom_bar(stat="identity", position="fill", aes(fill=variable)) +
  coord_flip() +
  scale_fill_manual(values = color16, name = "") +
  facet_grid(county~., scales="free") + 
  theme(legend.position = "bottom") +
  labs(title="Gut content composition from read count data across season and county\n(Harmonia DNA excluded)") +
  guides(fill=guide_legend(nrow=7, byrow=T))
dev.off()
```

PERMANOVA and Rarefaction curves on gut contents.

```{r joining guts reads}
ratio.guts = as.data.frame(reads.guts)
ratio.guts = ratio.guts[,!c(colnames(ratio.guts) %in% "Hemiptera")]
```

PERMANOVA analysis on the gut contents show a significant influence from all variables (season: p-value=0.0001; county: p-value=0.004); but not both (p-value=0.0883). It worths mentioning though that the residual values for these samples are all high; so we can assume that the influence of the low number of samples positive for prey taxa could have skewed the results.


```{r permanova guts reads}
info.guts = ratio.guts[,colnames(ratio.guts) %in% c("sample_id","sample_type","pcr","site","time","county")]
ratio.guts = ratio.guts[,!c(colnames(ratio.guts) %in% c("sample_id","sample_type","pcr","site","time","county"))]
ratio.guts = ratio.guts/rowSums(ratio.guts)

ratio.guts = ratio.guts[,!c(colnames(ratio.guts) %in% c("Harmonia_axyridis","Homo_sapiens"))]

ratio.guts = cbind(ratio.guts, info.guts)
rownames(ratio.guts) = ratio.guts$sample_id

ratio.guts = subset(ratio.guts, subset=rowSums(ratio.guts[,1:11]) > 0) #15

matrix.ratio.guts = ratio.guts[,!c(colnames(ratio.guts) %in% c("sample_id","sample_type","pcr","site","time","county"))]
matrix.ratio.guts$Dinocampus_coccinellae = 0

matrix.ratio.guts[,][matrix.ratio.guts[] > 0] = 1

fact_county.g = matrix(c(ratio.guts$county), ncol=1) 
fact_time.g = matrix(c(ratio.guts$time), ncol=1)
fact_all.g = matrix(c(ratio.guts$county, ratio.guts$time), ncol=2)

dist.guts = vegdist(matrix.ratio.guts, method="euclidean")

adonis(dist.guts~fact_county.g*fact_time.g, permutations = 9999, method="euclidean")
```

```{r metamds guts}
meta.guts = metaMDS(dist.guts, trymax=200, distance="euclidean")
```

```{r permanova guts plot}
png("R_images/NMDS_gut_contents.png", height=4000, width=4000, units="px", res=300)
plot(meta.guts,type="n")
points(meta.guts, select=which(fact_county.g[,1]=="Oxfordshire" & fact_time.g[,1]=="May"),col="black",pch=1)
points(meta.guts, select=which(fact_county.g[,1]=="Yorkshire" & fact_time.g[,1]=="May"),col="blue",pch=22)
points(meta.guts, select=which(fact_county.g[,1]=="Oxfordshire" & fact_time.g[,1]=="Oct"),col="black",pch=16)
points(meta.guts, select=which(fact_county.g[,1]=="Yorkshire" & fact_time.g[,1]=="Oct"),col="blue",pch=15)
ordispider(meta.guts,group=fact_county.g[,1], col=c("green", "green4"),label=T)
ordiellipse(meta.guts,group=fact_time.g[,1],kind="ehull", lty=c(1,3),label=T,lwd=2)
text(0.7,-0.5, scores(meta.guts$stress))
dev.off()
```

From the rarefaction curves of both species Richness and Shannon diversity extrapolated from the gut contents analysis it appear that more individuals might need to be collected in order to increase the number of detections from the gut contents.

```{r rarefaction curves on guts, include=T, echo=F}
abund.guts = as.data.frame(ratio.guts)
abund.guts[,1:15][abund.guts[,1:15] > 0] <- 1

abund.guts = as.data.frame(abund.guts[,!c(colnames(abund.guts) %in% c("sample_id","sample_type","pcr","site"))] %>% group_by(time, county) %>% summarise_all(funs(sum)))

raref.guts = t(abund.guts)
colnames(raref.guts) = c("Oxfordshire May","Yorkshire May","Oxfordshire Oct","Yorkshire Oct")
raref.guts = raref.guts[!c(rownames(raref.guts) %in% c("county","time")),]
raref.guts = as.data.frame(raref.guts)

raref.guts$`Oxfordshire May` = as.numeric(as.character(raref.guts$`Oxfordshire May`))
raref.guts$`Yorkshire May` = as.numeric(as.character(raref.guts$`Yorkshire May`))
raref.guts$`Oxfordshire Oct` = as.numeric(as.character(raref.guts$`Oxfordshire Oct`))
raref.guts$`Yorkshire Oct` = as.numeric(as.character(raref.guts$`Yorkshire Oct`))

raref.guts_inext = iNEXT(raref.guts, q=c(0,1), datatype = "abundance", se=T, endpoint=150)

strip_lab=c("0"="Species Richness","1"="Shannon Diversity")
ggiNEXT(raref.guts_inext) +
  labs(y="", x="Number detections") +
  theme(legend.position="bottom") +
  facet_grid(order~., labeller=as_labeller(strip_lab), scales="free")
```

# 6. Niche breadth

We calculate the niche breadth and niche overlap.

```{r diet/niche null model index}
niche.ratio = as.data.frame(ratio.guts)
niche.ratio[,1:15][niche.ratio[,1:15] > 0] <- 1

niche.breadth = as.data.frame(niche.ratio[,!c(colnames(niche.ratio) %in% c("sample_id","sample_type","pcr","site"))] %>% group_by(county, time) %>% summarise_all(list(sum)))

# We remove Dinocampus cocciinellae from the diet breadth
niche.breadth = niche.breadth[,!(colnames(niche.breadth) %in% "Dinocampus_coccinellae")]

rownames(niche.breadth) = paste(niche.breadth$county, niche.breadth$time)

niche.breadth = t(niche.breadth[,!c(colnames(niche.breadth) %in% c("county","time"))])

#niche.breadth[,1] = niche.breadth[,1]/107
#niche.breadth[,2] = niche.breadth[,2]/107

niche.width(niche.breadth, method="levins")
niche.width(niche.breadth, method="shannon")

(niche.width(niche.breadth, method="levins")[1]-1)/(nrow(niche.breadth)-1)
(niche.width(niche.breadth, method="levins")[2]-1)/(nrow(niche.breadth)-1)
(niche.width(niche.breadth, method="levins")[3]-1)/(nrow(niche.breadth)-1)
(niche.width(niche.breadth, method="levins")[4]-1)/(nrow(niche.breadth)-1)

niche.overlap(niche.breadth, method="pianka")
```

```{r niche overlap by season}
niche.breadth = as.data.frame(niche.ratio[,!c(colnames(niche.ratio) %in% c("sample_id","sample_type","pcr","site","county"))] %>% group_by(time) %>% summarise_all(list(sum)))

# We remove Dinocampus cocciinellae from the diet breadth
niche.breadth = niche.breadth[,!(colnames(niche.breadth) %in% "Dinocampus_coccinellae")]

rownames(niche.breadth) = paste("Ha",niche.breadth$time)

niche.breadth = t(niche.breadth[,!c(colnames(niche.breadth) %in% c("county","time"))])

#niche.breadth[,1] = niche.breadth[,1]/107
#niche.breadth[,2] = niche.breadth[,2]/107

niche.width(niche.breadth, method="levins")
niche.width(niche.breadth, method="shannon")

(niche.width(niche.breadth, method="levins")[1]-1)/(nrow(niche.breadth)-1)
(niche.width(niche.breadth, method="levins")[2]-1)/(nrow(niche.breadth)-1)
#(niche.width(niche.breadth, method="levins")[3]-1)/(nrow(niche.breadth)-1)
#(niche.width(niche.breadth, method="levins")[4]-1)/(nrow(niche.breadth)-1)

niche.overlap(niche.breadth, method="pianka")
```


# 7. Network of interactions

Tripartite network

```{r tripartite gut contents}
abund.guts$county = gsub("rd{1}|[a-z]{5}$", "", abund.guts$county)

rownames(abund.guts) = paste(abund.guts$county, abund.guts$time)
network.guts = abund.guts[,!c(colnames(abund.guts) %in% c("time","county"))]

network.guts = as.data.frame(t(network.guts))

rownames(network.guts)[10] <- "Chironomus"

web.ha = network.guts[!c(rownames(network.guts) %in% "Dinocampus_coccinellae"),]
web.dc = network.guts[c(rownames(network.guts) %in% "Dinocampus_coccinellae"),]

web.dc = as.data.frame(t(web.dc))
web.ha = as.matrix(web.ha)
web.dc = as.matrix(web.dc)

colnames(web.ha)
rownames(web.dc)
```

We set the order of prey by directly changing the order of rows into the network becuase the top level contains only one column, the parasite _Dinocampus coccinellae_ and changing the order using `sortweb(food.web, sort.order="seq", sequence=list(seq.higher=seq.high, seq.lower=seq.low)` is going to give errors in the top trophic level.

```{r order}
#"Betulaphis","Clethrobius_comes","Taeniothrips_inconsequens","Thripidae","Eucallipterus_tiliae",
web.ha = web.ha[c("Callipterinella","Trombidiformes_sp._BOLD:ACI3657","Euceraphis","Ocytata_pallipes","Drepanosiphum_platanoidis","Chrysomelidae","Chironomus","Ectopsocus_californicus","Psocoptera"),]

rownames(web.ha) = gsub("_|_[A-Z0-9:]{11,13}", " ", rownames(web.ha))
rownames(web.ha) = gsub("[a-z]{5,}\\s(?!sp)",".", rownames(web.ha), perl=T)
colnames(web.dc) = gsub("[a-z]{5,}_", ".", colnames(web.dc))
```

We can now plot the tripartite network.

```{r tripartite}
png("R_images/tripartite_network.png", width=6000, height=4000, res=300)
plotweb(web.dc, method="normal", labsize=2, y.lim=c(-0.5,2.7), x.lim=c(0,1.25), adj.high=c(0,-0.5), adj.low=c(0.5,-0.1), y.width.low=0.05, y.width.high=0.1, high.y=2.5, low.y=1.54, low.spacing=0.063, arrow="down", col.low="transparent", bor.col.low="transparent", col.interaction=c("orange","yellow","blue","red"), low.plot=T, empty=F)
plotweb(web.ha, method="normal", labsize=2,  text.rot=90, y.width.low=0.1, y.width.high=0.09,	high.y=1.5, low.y=0.5, low.lab.dis=0.05, high.lablength=0, high.spacing=0.07, low.spacing=0.02, col.high="transparent",  col.interaction=c("orange","yellow","blue","red"), add=T, empty=F)
dev.off()
```

# 8. Community composition.

We now investigate the arboreal communities associated with the _Harmonia axyridis_ sampled in both counties across the two separate seasons.

```{r join reads.comm plus summary, include=F}
reads.comms = full_join(oxf.comms, yor.comms)
reads.comms[is.na(reads.comms)] <- 0

reads.comms = reads.comms[,!c(colnames(reads.comms) %in% c("Harmonia_axyridis","Homo_sapiens"))]
reads.comms$Dinocampus_coccinellae = rowSums(reads.comms[,c(colnames(reads.comms) %in% c("Dinocampus","Dinocampus_coccinellae"))])
reads.comms = reads.comms[,!c(colnames(reads.comms) %in% "Dinocampus")]

reads.comms$Trombidiformes = rowSums(reads.comms[,c(colnames(reads.comms) %in% c("Trombidiformes_sp._BOLD:ACI3657","Trombidiformes"))])
reads.comms = reads.comms[,!c(colnames(reads.comms) %in% "Trombidiformes_sp._BOLD:ACI3657")]

reads.comms$Psocoptera = rowSums(reads.comms[,c(colnames(reads.comms) %in% c("Psocoptera_sp._BOLD:AAN8452","Psocoptera"))])
reads.comms = reads.comms[,!c(colnames(reads.comms) %in% "Psocoptera_sp._BOLD:AAN8452")]
```

```{r splitting communities into ranks, include=F}
info.comms = reads.comms[,c(colnames(reads.comms) %in% c("pcr","site","sample_id","sample_type","time","county"))]

sp.comms = reads.comms[,!c(colnames(reads.comms) %in% c("pcr","site","sample_id","sample_type","time","county"))][,grep("_", colnames(reads.comms[,!c(colnames(reads.comms) %in% c("pcr","site","sample_id","sample_type","time","county"))]))]

highrank.comms = reads.comms[,!c(colnames(reads.comms) %in% c("pcr","site","sample_id","sample_type","time","county"))][,grep("_", colnames(reads.comms[,!c(colnames(reads.comms) %in% c("pcr","site","sample_id","sample_type","time","county"))]), invert=T)]

melt.sp.comms = cbind(sp.comms, info.comms)
melt.rank.comms = cbind(highrank.comms, info.comms)
```

```{r community composition plots, include=F}
melt.comms = melt(melt.sp.comms)
melt.comms = subset(melt.comms, subset=melt.comms$value > 0)

#color19=c("#E69F00", "#56B4E9", "#CC79A7", "#009E73", "#F0E442", "#999999", "#0072B2", "#D55E00",  "#000000","#993300","#003333","#333300","#33FF33","#6633CC","#CC3300","#FF00CC","#CCFFCC", "#6633CC","#68228B")

png("R_images/arboreal_community_pres-abs.png", height=4000, width=4000, units="px", res=300)
ggplot(melt.comms, aes(time, value)) +
  geom_bar(stat="identity", position="fill", aes(fill=variable)) +
  facet_grid(.~county, scales="free") +
  theme(legend.position="none")
#  scale_fill_manual(values=color19) +
dev.off()
```

```{r join ratio.comms, include=F}
ratio.comms = as.data.frame(reads.comms)
```

PERMANOVA analysis (Number of permutations: 10000, Euclidean distance) on the community data show a strong significant influence of season (R2=0.36640, p-value=9.999e-05) on the investigated communities, rather than counties (R2=0.04647, p-value=0.2817); or both variables together (R2=0.06100, p-value=0.1396). This can be seen also in the NMDS plot with a small overlap between different season (ovals), but a strong overlap between county (webs).

```{r permanova communities, include=F}
info.comms = ratio.comms[,colnames(ratio.comms) %in% c("sample_id","sample_type","pcr","site","time","county")]
ratio.comms = ratio.comms[,!c(colnames(ratio.comms) %in% c("sample_id","sample_type","pcr","site","time","county"))]
ratio.comms = ratio.comms/rowSums(ratio.comms)

#ratio.comms[,][ratio.comms[,] > 0] <- 1
ratio.comms = cbind(ratio.comms, info.comms)
ratio.comms = ratio.comms[,!c(colnames(ratio.comms) %in% c("Harmonia_axyridis","Homo_sapiens"))]

adonis.ratio.com = ratio.comms[,!c(colnames(ratio.comms) %in% c("sample_id","sample_type","pcr","site","time","county"))]
```


```{r permanova communities on species, include=T}
ratio.sp.comms = sp.comms/rowSums(sp.comms)
ratio.sp.comms = cbind(ratio.sp.comms, info.comms)

adonis.sp.comm = ratio.sp.comms[,!c(colnames(ratio.sp.comms) %in% c("sample_id","sample_type","pcr","site","time","county"))]

fact_county = matrix(c(ratio.sp.comms$county),ncol=1) 
fact_time = matrix(c(ratio.sp.comms$time), ncol=1)
fact_all = matrix(c(ratio.sp.comms$county, ratio.sp.comms$time),ncol=2) 

adonis(adonis.sp.comm~fact_county*fact_time, permutations = 9999, method="jaccard")
```
```{r}
meta.comms = metaMDS(adonis.sp.comm, trymax=200, distance="jaccard")
```

```{r permanova low rank community plot, include=T, echo=F}
png("R_images/NMDS_arboreal_community.png", height=4000, width=4000, units="px", res=300)
plot(meta.comms,type="n", xlim=c(-1.5,1.5), ylim=c(-1.8,1.2))
points(meta.comms, select=which(fact_all[,1]=="Oxfordshire" & fact_all[,2]=="May"),col="black", pch=1)
points(meta.comms, select=which(fact_all[,1]=="Yorkshire" & fact_all[,2]=="May"),col="blue", pch=22)
points(meta.comms, select=which(fact_all[,1]=="Oxfordshire" & fact_all[,2]=="Oct"),col="black", pch=16)
points(meta.comms, select=which(fact_all[,1]=="Yorkshire" & fact_all[,2]=="Oct"),col="blue", pch=15)
ordispider(meta.comms,group=fact_all[,1], col=c("darkorange", "yellow"),label=T)
ordiellipse(meta.comms,group=fact_all[,2],kind="ehull", lty=c(1,3),label=T, lwd=2)
text(-0.4,-0.6, scores(meta.comms$stress))
dev.off()
```

Rarefaction curves based on taxa assigned at specie level, show that to reach the plateau of species richness in the communities, more samples should have been collected, or more sites should have been included; even though the diversity values appear to be much closer to the plateau.

```{r rarefaction curves on low ranks community, include=T, echo=F}
ratio.sp.comms[,!c(colnames(ratio.sp.comms) %in% c("sample_id","sample_type","pcr","site","county","time"))][ratio.sp.comms[,!c(colnames(ratio.sp.comms) %in% c("sample_id","sample_type","pcr","site","county","time"))] > 0] <- 1
abund.sp.comms = as.data.frame(ratio.sp.comms[,!c(colnames(ratio.sp.comms) %in% c("sample_id","sample_type","pcr","site"))] %>% group_by(time, county) %>% summarise_all(funs(sum)))

rare.sp.comms = t(abund.sp.comms)
colnames(rare.sp.comms) = c("Oxfordshire May","Yorkshire May","Oxfordshire Oct","Yorkshire Oct")
rare.sp.comms = rare.sp.comms[!c(rownames(rare.sp.comms) %in% c("county","time")),]
rare.sp.comms = as.data.frame(rare.sp.comms)

rare.sp.comms$`Oxfordshire May` = as.numeric(as.character(rare.sp.comms$`Oxfordshire May`))
rare.sp.comms$`Yorkshire May` = as.numeric(as.character(rare.sp.comms$`Yorkshire May`))
rare.sp.comms$`Oxfordshire Oct` = as.numeric(as.character(rare.sp.comms$`Oxfordshire Oct`))
rare.sp.comms$`Yorkshire Oct` = as.numeric(as.character(rare.sp.comms$`Yorkshire Oct`))

comms_inext = iNEXT(rare.sp.comms, q=c(0,1), datatype = "abundance", se=T, endpoint=100)

strip_labels=c("0"="Species Richness","1"="Shannon Diversity")
ggiNEXT(comms_inext) +
  labs(y = "", x="Number of samples") +
  theme(legend.position="bottom") +
  facet_grid(order~., labeller=as_labeller(strip_labels), scales="free") +
  ggtitle("Species level rarefaction curves on community")
```

Similarly to the taxa assigned to species level we ran the analysis also on the taxa assigned to higher ranks (family, order, etc.). and the results of both PERMANOVA (season: p-value=0.0008999; county: p-value=0.2576742; both variables: p-value=0.0391961) and rarefaction curves follow similar trends as per the species. Differences are in the rarefaction curves for the species richness, which appear to reach the plateau earlier than for the species. This might though be skewed by the lower number of taxa that were assigned to a higher rank.

```{r high rank communities, include=F}
ratio.highrank.comms = highrank.comms/rowSums(highrank.comms)
ratio.highrank.comms = cbind(ratio.highrank.comms, info.comms)
adonis.ratio.ranks = ratio.highrank.comms[,!c(colnames(ratio.highrank.comms) %in% c("sample_id","sample_type","pcr","site","time","county"))]

fact_county = matrix(c(ratio.highrank.comms$county), ncol=1) 
fact_time = matrix(c(ratio.highrank.comms$time), ncol=1)
fact_all = matrix(c(ratio.highrank.comms$county, ratio.highrank.comms$time), ncol=2) 

adonis(adonis.ratio.ranks~fact_county*fact_time, permutations = 9999, method="jaccard")
```

```{r}
meta.highcomms = metaMDS(adonis.ratio.ranks, trymax=200, distance="jaccard")
```


```{r permanova highranks taxa community plot, include=T, echo=F}
png("R_images/NMDS_highrank_community.png", height=4000, width=4000, units="px", res=300)
plot(meta.highcomms, type="n", ylim=c(-0.7,0.4))
points(meta.highcomms, select=which(fact_all[,1]=="Oxfordshire" & fact_all[,2]=="May"),col="black", pch=1)
points(meta.highcomms, select=which(fact_all[,1]=="Yorkshire" & fact_all[,2]=="May"),col="blue", pch=22)
points(meta.highcomms, select=which(fact_all[,1]=="Oxfordshire" & fact_all[,2]=="Oct"),col="black", pch=16)
points(meta.highcomms, select=which(fact_all[,1]=="Yorkshire" & fact_all[,2]=="Oct"),col="blue", pch=15)
ordispider(meta.highcomms,group=fact_all[,1], col=c("green", "green4"),label=T)
ordiellipse(meta.highcomms,group=fact_all[,2],kind="ehull", lty=c(1,3),label=T, lwd=2)
text(0.4, -0.6, scores(meta.highcomms$stress))
dev.off()
```

```{r rarefaction curves on community}
ratio.highrank.comms[,!c(colnames(ratio.highrank.comms) %in% c("sample_id","sample_type","pcr","site","county","time"))][ratio.highrank.comms[,!c(colnames(ratio.highrank.comms) %in% c("sample_id","sample_type","pcr","site","county","time"))] > 0] <- 1
abund.comms = as.data.frame(ratio.highrank.comms[,!c(colnames(ratio.highrank.comms) %in% c("sample_id","sample_type","pcr","site"))] %>% group_by(time, county) %>% summarise_all(funs(sum)))

rare.comms = t(abund.comms)
colnames(rare.comms) = c("Oxfordshire May","Yorkshire May","Oxfordshire Oct","Yorkshire Oct")
rare.comms = rare.comms[!c(rownames(rare.comms) %in% c("county","time")),]
rare.comms = as.data.frame(rare.comms)

rare.comms$`Oxfordshire May` = as.numeric(as.character(rare.comms$`Oxfordshire May`))
rare.comms$`Yorkshire May` = as.numeric(as.character(rare.comms$`Yorkshire May`))
rare.comms$`Oxfordshire Oct` = as.numeric(as.character(rare.comms$`Oxfordshire Oct`))
rare.comms$`Yorkshire Oct` = as.numeric(as.character(rare.comms$`Yorkshire Oct`))

comms_inext = iNEXT(rare.comms, q=c(0,1), datatype = "abundance", se=T, endpoint=100)

strip_labels=c("0"="Species Richness","1"="Shannon Diversity")
ggiNEXT(comms_inext) +
  labs(y = "", x="Number of samples") +
  theme(legend.position="bottom") +
  facet_grid(order~., labeller=as_labeller(strip_labels), scales="free") +
  ggtitle("High rank taxa rarefaction curves on community")
```

```{r permanova communities and gut contents}
ratio.guts[,!c(colnames(ratio.guts) %in% c("sample_id","sample_type","pcr","site","time","county"))][ratio.guts[,!c(colnames(ratio.guts) %in% c("sample_id","sample_type","pcr","site","time","county"))] > 0] <- 1

ratio.guts.comms = full_join(ratio.guts, ratio.sp.comms)

ratio.guts.comms[is.na(ratio.guts.comms)] = 0

# We remove Dinocampus coccinellae since it's a parasite and not directly part of the diet.
ratio.guts.comms[ratio.guts.comms$sample_type == "gut","Dinocampus_coccinellae"] = 0

adonis.guts.comms = ratio.guts.comms[,!c(colnames(ratio.guts.comms) %in% c("sample_id","sample_type","pcr","site","time","county"))]

fact_county = matrix(c(ratio.guts.comms$county),ncol=1) 
fact_time = matrix(c(ratio.guts.comms$time), ncol=1)
fact_type = matrix(c(ratio.guts.comms$sample_type), ncol=1)

adonis(adonis.guts.comms~fact_county*fact_time*fact_type, permutations = 9999, method="euclidean")
```

```{r}
meta.guts.comms = metaMDS(adonis.guts.comms, trymax=200, distance="euclidean")
```

```{r permanova gut contents and community plot, include=T, echo=F}
png("R_images/NMDS_guts_community.png", height=4000, width=4000, units="px", res=300)
plot(meta.guts.comms,type="n")#,xlim=c(-0.05,0.05), ylim=c(-0.01,0.01))
points(meta.guts.comms, select=which(fact_county[,1]=="Oxfordshire" & fact_time[,1]=="May"),col="black", pch=1)
points(meta.guts.comms, select=which(fact_county[,1]=="Yorkshire" & fact_time[,1]=="May"),col="blue", pch=22)
points(meta.guts.comms, select=which(fact_county[,1]=="Oxfordshire" & fact_time[,1]=="Oct"),col="black", pch=16)
points(meta.guts.comms, select=which(fact_county[,1]=="Yorkshire" & fact_time[,1]=="Oct"),col="blue", pch=15)
orditorp(meta.guts.comms, display="species")
ordispider(meta.guts.comms,group=fact_type[,1], col=c("darkorange", "yellow"),label=F)
ordiellipse(meta.guts.comms,group=fact_type[,1],kind="ehull", lty=c(1,3),label=T, lwd=2)
text(-0.4,-0.6, scores(meta.guts.comms$stress))
dev.off()
```

# 9. Generating data summary table

This last chunk is only to create the base for the summary table presented in supplementary informaton. This uses the presence/absence data for the gut contents and the DNA bulk samples. The table was then expanded externally to R to include more summarising informations.

```{r generating data summary table}
colnames(niche.ratio)
guts.summary = as.data.frame(niche.ratio[,!c(colnames(niche.ratio) %in% c("sample_id"))] %>% group_by(sample_type,time,county,site,pcr) %>% summarise_all(list(sum)))

colnames(ratio.sp.comms)
comms.summary = as.data.frame(ratio.sp.comms[,!c(colnames(ratio.sp.comms) %in% c("sample_id"))] %>% group_by(sample_type,time,county,site,pcr) %>% summarise_all(list(sum)))

full.data = full_join(guts.summary, comms.summary)

# Splitting the metadata from data, so that we can remove the NAs into the data set.
full.info = full.data[,c(colnames(full.data) %in% c("sample_type","time","county","site","pcr"))]
full.data = full.data[,!c(colnames(full.data) %in% c("sample_type","time","county","site","pcr"))]
full.data[is.na(full.data)] <- 0

# Binding the frame together using specific order of the metadata
final.data = bind_cols(full.info[,c(1,2,3,4)], full.data)

# Transposing and saving the data frame, we will rearrange the order of the metadata later
final.data = t(final.data)

#Saving final output
write.csv(final.data, "R_outputs/Data_summary_table.csv")
```

# 10. Publication plot

```{r publication blocking primers plot}
full_blk_melt$pcr = factor(full_blk_melt$pcr, levels=c("no-blocking primers","blocking primers"))
blk_a = ggplot(full_blk_melt,(aes(specie_type, value))) +
  geom_boxplot(aes(fill=specie_type), outlier.shape = NA) +
  geom_jitter(alpha=0.5, size=4) +
  labs(x="", y="Normalised DNA reads") +
  theme_bw() +
  facet_grid(.~pcr) +
  theme(legend.position="none", axis.text.x=element_blank(), axis.ticks.x=element_blank(), axis.text=element_text(size=15), axis.title=element_text(size=15), strip.text=element_text(size=15), plot.margin=unit(c(0,0.25,0,0.25), unit="cm"))

full_blk_melt$pcr = factor(full_blk_melt$pcr, levels=c("no-blocking primers","blocking primers"))
blk_b = ggplot(full_blk_melt,(aes(specie_type, log10(value)))) +
  geom_boxplot(aes(fill=specie_type), outlier.shape = NA) +
  geom_jitter(alpha=0.5, size=4) +
  labs(x="", y="Log10(Normalised DNA reads)") +
  theme_bw() +
  facet_grid(.~pcr) +
  theme(legend.position="right", legend.title=element_blank(), legend.text=element_text(size=20), axis.text.x=element_blank(), axis.ticks.x=element_blank(), axis.text=element_text(size=15), axis.title=element_text(size=15), strip.text=element_text(size=15), plot.margin=unit(c(0,0,0,0.6), unit="cm"))

png("R_images/publication_blocking-primers.png", height=4000, width=4000, units="px", res=300)
plot_grid(blk_a, blk_b, labels=c("a)","b)"), ncol=2, nrow=1, label_size=25, label_fontface=1)
dev.off()
```

```{r species names}
heat_blk_melt$specie = gsub("[a-z]{7,}_", ".", heat_blk_melt$specie)
heat_blk_melt$pcr = as.character(heat_blk_melt$pcr)

heat_blk_melt[(heat_blk_melt$pcr == "no-blocking primers"), "pcr"] = "no-blk"
heat_blk_melt[(heat_blk_melt$pcr == "blocking primers"),"pcr"] = "blk"
```

```{r publication heatmap blocking primers}
heat_a<-ggplot(heat_blk_melt[heat_blk_melt$specie %in% c("D.platanoidis","P.viburni"),],(aes(pcr, variable))) +
  geom_tile(aes(fill=value), linetype=1, color="gray70", size=0.2) +
  scale_fill_gradientn(colors=c("white","grey87","orange","blue"), values=c(0,0.01,0.5,1), na.value="black", name="Normalised\nDNA reads\n") + 
  #scale_fill_gradient2(low="gray87", high="blue", mid="orange", midpoint=40, na.value="green", limits=c(1,70), name="Normalised\nDNA reads\n") +
  scale_x_discrete(limits=c("no-blk","blk")) +
  theme_classic() +
  facet_grid(.~specie) + labs(x="", y="") +
  theme(legend.position="right", legend.text=element_text(size=15), axis.text.x=element_text(size=15), axis.ticks.x=element_blank(), axis.text=element_text(size=15), axis.title=element_text(size=15), strip.text=element_text(size=12), plot.margin=unit(c(0,0,0,0), unit="cm"))

heat_b<-ggplot(heat_blk_melt[heat_blk_melt$specie %in% c("Chrysomelidae","E.betulae"),],(aes(pcr, variable))) +
  geom_tile(aes(fill=value), linetype=1, color="gray70", size=0.2) +
  scale_fill_gradientn(colors=c("white","grey87","orange","blue"), values=c(0,0.01,0.5,1), na.value="black", name="Normalised\nDNA reads\n") + 
  #scale_fill_gradient2(low="gray87", high="blue", mid="orange", midpoint=2000, na.value="transparent", limits=c(1,4000), name="Normalised\nDNA reads\n") + 
  scale_x_discrete(limits=c("no-blk","blk")) +
  theme_classic() +
  facet_grid(.~specie) + labs(x="", y="") +
  theme(legend.position="right", legend.text=element_text(size=15), axis.text.x=element_text(size=15), axis.ticks.x=element_blank(), axis.text.y = element_blank(), axis.text=element_text(size=15), axis.title=element_text(size=15), strip.text=element_text(size=12), plot.margin=unit(c(0,0,0,0), unit="cm"))

heat_c<-ggplot(heat_blk_melt[heat_blk_melt$specie_type %in% "predator",],(aes(pcr, variable))) +
  geom_tile(aes(fill=value), linetype=1, color="gray70", size=0.2) +
  scale_fill_gradientn(colors=c("grey87","orange","blue"), values=c(0,0.5,1), na.value="black", name="Normalised\nDNA reads\n") + 
  #scale_fill_gradient2(low="gray87", high="blue", mid="orange", midpoint=35000, na.value="transparent", limits=c(1,70000), name="Normalised\nDNA reads\n") +
  scale_x_discrete(limits=c("no-blk","blk")) +
  theme_classic() +
  facet_grid(.~specie) + labs(x="", y="") +
  theme(legend.position="right", legend.text=element_text(size=15), axis.text.x=element_text(size=15), axis.ticks.x=element_blank(), axis.text.y = element_blank(), axis.text=element_text(size=15), axis.title=element_text(size=15), strip.text=element_text(size=12), plot.margin=unit(c(0,0,0,0), unit="cm"))

png("R_images/publication_heatmap_block-primers.png", height=3000, width=6000, units="px", res=300)
grid.arrange(heat_a, heat_b, heat_c, ncol=3, nrow=1, left="Sample ID")
dev.off()
```


```{r publication NMDS guts}
png("R_images/publication_NMDS_guts.png", height=4000, width=4000, units="px", res=300)

plot(meta.guts,type="n", xlim=c(-1,1.2), ylim=c(-1,1), cex.axis=2, cex.lab=1.5)
points(meta.guts, select=which(fact_county.g[,1]=="Oxfordshire" & fact_time.g[,1]=="May"),col="black",pch=1, cex=2)
points(meta.guts, select=which(fact_county.g[,1]=="Yorkshire" & fact_time.g[,1]=="May"),col="blue",pch=22, cex=2)
points(meta.guts, select=which(fact_county.g[,1]=="Oxfordshire" & fact_time.g[,1]=="Oct"),col="black",pch=16, cex=2)
points(meta.guts, select=which(fact_county.g[,1]=="Yorkshire" & fact_time.g[,1]=="Oct"),col="blue",pch=15, cex=2)
ordispider(meta.guts,group=fact_county.g[,1],col=c("green", "green4"),label=T, cex=1.5)
ordiellipse(meta.guts,group=fact_time.g[,1],kind="ehull", lty=c(1,3),label=F,lwd=1)
text(-0.8,-0.5,label="May", cex=2.5)
text(-0.8,0.5,label="Oct", cex=2.5)
text(0.7,-0.9, label=paste0("Stress: ", round(meta.guts$stress, 4)), cex=1.5)
text(0.5,0.9,label="gut contents", cex=2.5)

legend("bottomleft", legend=c("Oxfordshire - May","Oxfordshire - Oct","Yorkshire - May","Yorkshire - Oct"), col=c("black","black","blue","blue"), pch=c(1,16,22,15), cex=2, ncol=2, xpd=T)
dev.off()
```

```{r publication NMDS comms and comms vs guts}
png("R_images/publication_NMDS_comms_guts_plots.png", height=4000, width=4000, units="px", res=300)
lay.matrix = matrix(c(1,1,0,0,1,1,4,0,2,2,3,0,2,2,2,0), nrow=4, ncol=4, byrow=T)
layout(lay.matrix)

plot(meta.comms,type="n", xlim=c(-3,3), ylim=c(-4,2.5), cex.axis=2, cex.lab=1.5)
points(meta.comms, select=which(fact_all[,1]=="Oxfordshire" & fact_all[,2]=="May"),col="black", pch=1, cex=2)
points(meta.comms, select=which(fact_all[,1]=="Yorkshire" & fact_all[,2]=="May"),col="blue", pch=22, cex=2)
points(meta.comms, select=which(fact_all[,1]=="Oxfordshire" & fact_all[,2]=="Oct"),col="black", pch=16, cex=2)
points(meta.comms, select=which(fact_all[,1]=="Yorkshire" & fact_all[,2]=="Oct"),col="blue", pch=15, cex=2)
ordispider(meta.comms,group=fact_all[,1],col=c("darkorange", "yellow"),label=T, cex=1.5)
ordiellipse(meta.comms,group=fact_all[,2],kind="ehull", lty=c(1,3),label=F, lwd=1)
text(-3,-0.5,label="May", cex=2.5)
text(3,2,label="Oct", cex=2.5)
text(2.5,-3.7, label=paste0("Stress: ", round(meta.comms$stress, 4)), cex=1.5)
text(-3,2.3, label="a)", cex=3)
text(0,2.3,label="community", cex=2.5)

plot(meta.guts.comms,type="n", xlim=c(-2,5), ylim=c(-2.5,2.5), cex.axis=2, cex.lab=1.5)
points(meta.guts.comms, select=which(fact_county[,1]=="Oxfordshire" & fact_time[,1]=="May"),col="black", pch=1, cex=2)
points(meta.guts.comms, select=which(fact_county[,1]=="Yorkshire" & fact_time[,1]=="May"),col="blue", pch=22, cex=2)
points(meta.guts.comms, select=which(fact_county[,1]=="Oxfordshire" & fact_time[,1]=="Oct"),col="black", pch=16, cex=2)
points(meta.guts.comms, select=which(fact_county[,1]=="Yorkshire" & fact_time[,1]=="Oct"),col="blue", pch=15, cex=2)
ordispider(meta.guts.comms,group=fact_type[,1],col=c("darkorange", "green4"),label=F, cex=1.5)
ordiellipse(meta.guts.comms,group=fact_time[,1],kind="ehull", lty=c(1,3), label=F, lwd=2)
text(3,-2.4, label=paste0("Stress: ", round(meta.guts.comms$stress, 4)), cex=1.5)
text(-2,2.2, label="b)", cex=3)
text(2,2,label="Oct", cex=2.5)
text(-2,0,label="May", cex=2.5)

plot(meta.guts.comms,type="n", xlim=c(-0.2,0.1), ylim=c(-0.2,0.3), xlab="", ylab="", cex.axis=2)
ordispider(meta.guts.comms,group=fact_type[,1],col=c("darkorange", "green4"),label=T, cex=1.5)
ordiellipse(meta.guts.comms,group=fact_time[,1],kind="ehull",lty=c(1,3),label=F, lwd=2)
points(meta.guts.comms, select=which(fact_county[,1]=="Oxfordshire" & fact_time[,1]=="May"),col="black", pch=1, cex=2)
points(meta.guts.comms, select=which(fact_county[,1]=="Yorkshire" & fact_time[,1]=="May"),col="blue", pch=22, cex=2)
points(meta.guts.comms, select=which(fact_county[,1]=="Oxfordshire" & fact_time[,1]=="Oct"),col="black", pch=16, cex=2)
points(meta.guts.comms, select=which(fact_county[,1]=="Yorkshire" & fact_time[,1]=="Oct"),col="blue", pch=15, cex=2)

# Common legend
plot.new()
legend("bottom", legend=c("Oxfordshire - May","Oxfordshire - Oct","Yorkshire - May","Yorkshire - Oct"), col=c("black","black","blue","blue"), pch=c(1,16,22,15), cex=2, ncol=1)
dev.off()
```

```{r publication tripartite plot}
name.legend=c("Hemiptera","Trombidiformes","Diptera","Coleoptera","Psocoptera")
col.legend=c("cyan3","orchid","tan4","darkolivegreen4","khaki")
col.prey=c("cyan3","orchid","cyan3","tan4","cyan3","darkolivegreen4","tan4","khaki","khaki")

web.ha
```

```{r tripartite network}
png("R_images/publication_tripartite_network.png", width=6000, height=4000, res=300)
plotweb(web.dc, method="normal", labsize=3, y.lim=c(-0.5,2.7), x.lim=c(0,1.25), adj.high=c(0,-0.5), adj.low=c(0.5,-0.1), y.width.low=0.05, y.width.high=0.1, high.y=2.5, high.xoff = 0.1, low.xoff = 0.1, low.y=1.54, low.spacing=0.03, arrow="down", col.low="transparent", bor.col.low="transparent", col.interaction=c("orange","yellow","blue","red"), low.plot=T, empty=F)
plotweb(web.ha, method="normal", labsize=3,  text.rot=90, y.width.low=0.1, y.width.high=0.09,	high.y=1.5, low.y=0.5, low.lab.dis=0.05, high.lablength=0, high.spacing=0.077, low.spacing=0.03, col.high="transparent",  col.interaction=c("orange","yellow","blue","red"), col.low=col.prey, add=T, empty=F)
#web.dc interactions
text(0.2,2.3, paste("N =", web.dc[1,1]), cex=1.5)
text(0.375,2.3, paste("N =", web.dc[2,1]), cex=1.5)
text(0.7,2.3, paste("N =", web.dc[3,1]), cex=1.5)
# web.ha interactions
text(0.01,1.2, "*", cex=3)
text(0.035,1.2, "*", cex=3)
text(0.15,1.2, paste("N =", web.ha[3,1]), cex=1.5) 
text(0.5,1.2, paste("N =", web.ha[3,2]), cex=1.5)
text(0.755,1.2, "*", cex=3)
text(0.785,1.2, "*", cex=3)
text(0.83,1.2, "*", cex=3)
text(0.91,1.2, paste("N =", web.ha[3,4]), cex=1.5)
text(1.035,1.2, "*", cex=3)
text(1.065,1.2, "*", cex=3)
text(1.11,1.2, paste("N =", web.ha[8,4]), cex=1.5)
text(1.2,1.2, paste("N =", web.ha[9,4]), cex=1.5)
legend("bottom", legend=name.legend, fill=col.legend, bty="n", border="black",ncol=6, xpd=T, cex=1.5)
dev.off()
```

```{r publication barplot simplified}
#unique(melt.comms$variable)
melt.guts = melt.guts[-(grep("Dinocampus_coccinellae", melt.guts$variable)),]

melt.comms.guts = full_join(melt.comms, melt.guts)

melt.comms.guts$order = "NA"
melt.comms.guts[grep("Acanthosoma|Anthocoris|Coreus|Deraeocoris|Elasmucha|Neolygus|Palomena|Pentatoma|Psallus", melt.comms.guts$variable),"order"] = "Hemiptera - Heteroptera"
melt.comms.guts[grep("Agathomyia|Chironomus|Lonchoptera|Ocytata|Schwenckfeldina", melt.comms.guts$variable), "order"] = "Diptera"
melt.comms.guts[grep("Agriopis|Amphipyra|Bucculatrix|Cameraria|Orthosia|Pandemis|Tiliacea", melt.comms.guts$variable), "order"] = "Lepidoptera"
melt.comms.guts[grep("Anelosimus|Clubiona|Hylyphantes|Metellina|Oligolophus|Paidiscura|Paroligolophus|Philodromus|Tenuiphantes|Tetragnatha|Theridion|Trombidiformes", melt.comms.guts$variable), "order"] = "Arachnida"
melt.comms.guts[grep("Aphthona|Cantharis|Chrysomelidae|Enicmus|Phyllopertha|Polydrusus|Rhagonycha|Rhyzobius", melt.comms.guts$variable), "order"] = "Coleoptera"
melt.comms.guts[grep("Cacopsylla|Callipterinella|Drepanosiphum|Eucallipterus|Euceraphis|Macrosiphum|Psyllopsis|Rhopalosiphum", melt.comms.guts$variable), "order"] = "Hemiptera - Aphidomorpha"
melt.comms.guts[grep("Cepaea", melt.comms.guts$variable), "order"] = "Gastropoda"
melt.comms.guts[grep("Chrysoperla|Hemerobius", melt.comms.guts$variable), "order"] = "Neuroptera"
melt.comms.guts[grep("Entomobrya|Sminthurus", melt.comms.guts$variable), "order"] = "Collembola"
melt.comms.guts[grep("Glyphotaelius", melt.comms.guts$variable), "order"] = "Trichoptera"
melt.comms.guts[grep("Lasius|Pristiphora|Trioxys", melt.comms.guts$variable), "order"] = "Hymenoptera"
melt.comms.guts[grep("Panorpa", melt.comms.guts$variable), "order"] = "Mecoptera"
melt.comms.guts[grep("Tachypodoiulus", melt.comms.guts$variable), "order"] = "Myriapoda"
melt.comms.guts[grep("Ectopsocus|Graphopsocus|Valenzuela|Psocoptera", melt.comms.guts$variable), "order"] = "Psocoptera"
```

```{r publication barplot}
col.barplot = c("orange","orchid","grey60","yellow","cyan","green","tan4","gold3","red","blue","khaki","aquamarine","olivedrab","black")

png("R_images/publication_barplot_composition.png", height=4000, width=4000, units="px", res=300)
ggplot(melt.comms.guts, aes(sample_type,value)) +
  geom_bar(stat="identity", position="fill", aes(fill=order)) +
  scale_y_continuous(expand=c(0,0)) +
  scale_fill_manual(values=col.barplot, name = "") +
  xlab("") +
  ylab("DNA reads ratio") +
  coord_flip() +
  facet_grid(county*time~.) +
  theme_bw() +
  theme(legend.position = "right", axis.text = element_text(size = 15), strip.text = element_text(size=15), 
        axis.title = element_text(size =15),legend.text = element_text(size=15))
dev.off()  
```

```{r log session info}
sink("log_session_analysis.txt")
sessionInfo()
citation()
sink()
```

```{r packages citation}
sink("packages_citations.txt")
for (p in pack.list){
 print(citation(p))
}
sink()
```

```{r end}
print("end")
```